src/main/java/io/nats/client/impl/NatsMessageConsumer.java

```diff
@@ -52,43 +52,50 @@ class NatsMessageConsumer extends NatsMessageConsumerBase implements PullManager
     @Override
     public void heartbeatError() {
         try {
-            // just close the current sub and make another one.
-            // this could go on endlessly - unless the user had called stop
             if (stopped.get()) {
-                finishAndClose();
+                finishAndShutdownSub();
             }
             else {
-                lenientClose();
+                shutdownSub();
                 doSub();
             }
         }
         catch (JetStreamApiException | IOException e) {
-            pmm.resetTracking();
-            pmm.initOrResetHeartbeatTimer();
+            setupHbAlarmToTrigger();
         }
     }
 
     void doSub() throws JetStreamApiException, IOException {
         MessageHandler mh = userMessageHandler == null ? null : msg -> {
             userMessageHandler.onMessage(msg);
             if (stopped.get() && pmm.noMorePending()) {
-                finishAndClose();
+                finished.set(true);
             }
         };
-        super.initSub(subscriptionMaker.subscribe(mh, userDispatcher, pmm, null));
-        repull();
-        stopped.set(false);
-        finished.set(false);
+        try {
+            stopped.set(false);
+            finished.set(false);
+            super.initSub(subscriptionMaker.subscribe(mh, userDispatcher, pmm, null));
+            repull();
```

Comment:
this also matches before, although I did move the setting of the flags before instead of after